superset-frontend/src/SqlLab/components/AceEditorWrapper/useKeywords.ts

```diff
@@ -0,0 +1,200 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+import { useEffect, useMemo, useRef } from 'react';
+import { useSelector, useDispatch, shallowEqual, useStore } from 'react-redux';
+import { t } from '@superset-ui/core';
+
+import { Editor } from 'src/components/AsyncAceEditor';
+import sqlKeywords from 'src/SqlLab/utils/sqlKeywords';
+import { addTable, addDangerToast } from 'src/SqlLab/actions/sqlLab';
+import {
+  SCHEMA_AUTOCOMPLETE_SCORE,
+  TABLE_AUTOCOMPLETE_SCORE,
+  COLUMN_AUTOCOMPLETE_SCORE,
+  SQL_FUNCTIONS_AUTOCOMPLETE_SCORE,
+} from 'src/SqlLab/constants';
+import {
+  useSchemas,
+  useTables,
+  tableEndpoints,
+  skipToken,
+} from 'src/hooks/apiResources';
+import { api } from 'src/hooks/apiResources/queryApi';
+import { useDatabaseFunctionsQuery } from 'src/hooks/apiResources/databaseFunctions';
+import useEffectEvent from 'src/hooks/useEffectEvent';
+import { SqlLabRootState } from 'src/SqlLab/types';
+
+type Params = {
+  queryEditorId: string | number;
+  dbId?: string | number;
+  schema?: string;
+};
+
+const EMPTY_LIST = [] as typeof sqlKeywords;
+
+export function useKeywords(
+  { queryEditorId, dbId, schema }: Params,
+  skip = false,
+) {
+  const dispatch = useDispatch();
+  const hasFetchedKeywords = useRef(false);
+  // skipFetch is used to prevent re-evaluating memoized keywords
+  // due to updated api results by skip flag
+  const skipFetch = hasFetchedKeywords && skip;
+  const { data: schemaOptions } = useSchemas({
```

Comment:
I'm not sure if I understood. I'm assuming this hook is independent of the `DatabaseSelector`. It's calling `useSchemas` from `src/hooks/apiResources` which might fail right? If I use this hook in another component, shouldn't the exception be handled?

Reply:
You're right. This hook is  independent of the DatabaseSelector. However, it shared the query data from DatabaseSelector by using shared `useSchemas/useTables` hooks. If this hook also add the error handler (which shows the error toast), the same error message pops up twice because that error handler is already implemented in `DatabaseSelect` which is used in the left sidebar layout.

Reply:
Why the error handler in `DatabaseSelector`? There's a benefit of covering all error situations in that component such like [new dataset left menu refactor](https://github.com/apache/superset/pull/24599) doesn't need to add extra error handling logic for that integration.

Reply:
To minimize the confusion, let me change `useQuery` over `useQueryState` which means only sharing the query data without triggering fetches. It will only useQuery for database_functions since it's the only item that requires fetch for additional autocomplete data

Reply:
As we discussed offline, let's improve the error handling in a follow-up PR as it may require a more detailed context.